### PR TITLE
#896 add optimized byte array serializer in default config

### PR DESCRIPTION
**Pull request by lorban**
Comments are welcome but I'd like to wait for the perf results before this gets merged in.

### CODE REVIEW

File: impl/src/main/java/org/ehcache/impl/serialization/ByteArraySerializer.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    binary.get(bytes);
+    return bytes;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean equals(byte[] object, ByteBuffer binary) throws ClassNotFoundException, SerializerException {
+    return Arrays.equals(read(binary), object);
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/serialization/ByteArraySerializer.java]: This isn't the equals contract of arrays. (also wouldn't it be faster to do `binary.equals(serialize(object))`?)

Comment (lorban) [inline impl/src/main/java/org/ehcache/impl/serialization/ByteArraySerializer.java]: Two remarks:

1) replacing `Arrays.equals(read(binary), object)` with `binary.equals(serialize(object))` would indeed be faster but the latter doesn't 'consume' the bytes of the byte buffer (position remains unchanged). All the other serializers do but that isn't documented. Should this impl also consume the bytes or not (by updating the position in your proposed impl)? And should this be documented or not?

2) I know that this isn't the real `byte[].equals()` contract, but without this 'improvement', atomic operations and the XAStore won't work. And somewhat unrelated, Hibernate stores `Object[]` values IIRC, which means we might need to add more array serializers to make it work with XA caches.

File: impl/src/main/java/org/ehcache/impl/serialization/ByteArraySerializer.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    return bytes;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean equals(byte[] object, ByteBuffer binary) throws ClassNotFoundException, SerializerException {
+    boolean equals = binary.equals(serialize(object));
+    binary.position(binary.remaining());
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/serialization/ByteArraySerializer.java]: I think you want `binary.position(binary.limit())` (and I still think this is wrong)

**Comment by lorban** [comment]
test this please

**Comment by AbfrmBlr** [comment]
test this pleasea

**Comment by ljacomet** [comment]
Any news on perf results? @lorban @skbansal

**Pull request merged**